Makefile

```diff
@@ -1,29 +1,153 @@
+LDFLAGS =
 
-.PHONY: clean
-clean: 
-	rm $(TARGET_BPF) $(VMLINUX_H)
+# golang
+
+CGO_CFLAGS_STATIC = "-I$(abspath $(OUTPUT))"
+CGO_LDFLAGS_STATIC = "-lelf -lz $(LIBBPF_OBJ)"
+
+CGO_CFGLAGS_DYN = "-I. -I/usr/include/"
+CGO_LDFLAGS_DYN = "-lelf -lz -lbpf"
+
+# default == shared lib from OS package
+
+all: libbpfgo-dynamic
+install: libbpfgo-dynamic-install
+test: libbpfgo-dynamic-test
+
+# libbpfgo test object
+
+libbpfgo-test-bpf:
+	$(MAKE) -C $(SELFTEST)/build
+
+libbpfgo-test-bpf-clean:
+	$(MAKE) -C $(SELFTEST)/build clean
+
+# libbpf: shared
+
+libbpfgo-dynamic: vmlinuxh outputlibdir
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_DYN) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_DYN) \
+	   go build .
+
+libbpfgo-dynamic-install:
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_DYN) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_DYN) \
+	   go install .
+
+libbpfgo-dynamic-test: libbpfgo-test-bpf
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_DYN) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_DYN) \
+	   sudo -E go test .
+
+# libbpf: static
+
+libbpfgo-static: $(LIBBPF_OBJ) vmlinuxh
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_STATIC) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_STATIC) \
+	   go build .
+
+libbpfgo-static-install:
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_STATIC) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_STATIC) \
+	   go install .
+
+libbpfgo-static-test: libbpfgo-test-bpf
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_STATIC) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_STATIC) \
+	   sudo -E go test .
+
+# vmlinux header file
+
+vmlinuxh: outputdir
+	@if [ ! -f $(VMLINUX_H) ]; then \
+		$(BPFTOOL) btf dump file /sys/kernel/btf/vmlinux format c > $(VMLINUX_H); \
+	fi
+
+# static libbpf generation for the git submodule
+
+$(LIBBPF_OBJ): $(wildcard $(LIBBPF_SRC)/*.[ch]) | outputlibdir
+	CC="$(CC)" CFLAGS="$(CFLAGS)" LD_FLAGS="$(LDFLAGS)" \
+	   $(MAKE) -C $(LIBBPF_SRC) \
+		BUILD_STATIC_ONLY=1 \
+		OBJDIR=$(LIBBPF_OBJDIR) \
+		DESTDIR=$(LIBBPF_DESTDIR) \
+		INCLUDEDIR= LIBDIR= UAPIDIR= install
+
+# selftests
+
+SELFTESTS = $(shell find $(SELFTEST) -mindepth 1 -maxdepth 1 -type d ! -name 'common' ! -name 'build')
+
+define FOREACH
+	SELFTESTERR=0; \
+	for DIR in $(SELFTESTS); do \
+		$(MAKE) -j8 -C $$DIR $(1) || SELFTESTERR=1; \
+	done; \
+	if [ $$SELFTESTERR -eq 1 ]; then \
+		exit 1; \
+	fi
+endef
+
+selftest: selftest-static
+selftest-run: selftest-static-run
+
+.PHONY: selftest
+.PHONY: selftest-static
+.PHONY: selftest-dynamic
+.PHONY: selftest-clean
+
+.PHONY: selftest-run
+.PHONY: selftest-static-run
+.PHONY: selftest-dynamic-run
+
+selftest-static:
+	$(call FOREACH, main-static)
+selftest-dynamic:
+	$(call FOREACH, main-dynamic)
+selftest-static-run:
+	$(call FOREACH, run-static)
+selftest-dynamic-run:
+	$(call FOREACH, run-dynamic)
+
+selftest-clean:
+	$(call FOREACH, clean)
+
+# output
+
+.PHONY: $(OUTPUT)
+.PHONY: outputdir
+.PHONY: outputlibdir
+
+outputdir:
+	mkdir -p $(OUTPUT)
+outputlibdir:
+	mkdir -p $(OUTPUT)/libbpf
+
+# cleanup
+
+clean: selftest-clean libbpfgo-test-bpf-clean
```

Comment:
`go build` always keeps them cached. I read somewhere we should let go to manage the cache. We could always force the full build (bypassing the cache with `go build -a`).

> To clean local caches I'm currently doing `go clean -cache ; go clean -modcache ; go clean -testcache`.

Reply:
I meant more about the install target (which I'm not sure about it's purpose)

Reply:
yep, I answered this one bellow. Install isn't really needed if we are not providing binaries. It would install under GOPATH/bin/ but we're only caching objects when doing 'go build' (thus my answer). Marking this as resolved since I removed 'make install, make libbpfgo-{static,dynamic}-install, etc...'.